src/Nethermind/Nethermind.Shutter/ShutterTxLoader.cs

```diff
@@ -0,0 +1,242 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Nethermind.Core;
+using Nethermind.Core.Specs;
+using Nethermind.Crypto;
+using Nethermind.Blockchain.Find;
+using Nethermind.Int256;
+using Nethermind.Serialization.Rlp;
+using Nethermind.Logging;
+using Nethermind.Shutter.Contracts;
+using Nethermind.Shutter.Config;
+using Nethermind.Core.Collections;
+using Nethermind.Core.Extensions;
+using System.IO;
+using Nethermind.Abi;
+
+namespace Nethermind.Shutter;
+
+using G1 = Bls.P1;
+
+public class ShutterTxLoader(
+    ILogFinder logFinder,
+    IShutterConfig cfg,
+    ShutterTime time,
+    ISpecProvider specProvider,
+    IEthereumEcdsa ecdsa,
+    IAbiEncoder abiEncoder,
+    ILogManager logManager)
+{
+    private readonly ShutterLogScanner _logScanner = new(
+                new(new Address(cfg.SequencerContractAddress!)),
+                logFinder,
+                logManager,
+                abiEncoder);
+
+    private readonly ShutterEventQueue _events = new(cfg.EncryptedGasLimit, logManager);
+    private ulong _txPointer = ulong.MaxValue;
+    private bool _loadFromReceipts = false;
+    private readonly ShutterTxFilter _txFilter = new(specProvider, logManager);
+    private readonly ILogger _logger = logManager.GetClassLogger();
+
+    public ShutterTransactions LoadTransactions(Block? head, BlockHeader parentHeader, IShutterKeyValidator.ValidatedKeyArgs keys)
+    {
+        var sequencedTransactions = GetNextTransactions(keys.Eon, keys.TxPointer, head?.Number ?? 0).ToList();
+
+        long offset = time.GetCurrentOffsetMs(keys.Slot);
+        Metrics.KeysReceivedTimeOffset = offset;
+        string offsetText = offset < 0 ? $"{-offset}ms before" : $"{offset}ms after";
+        _logger.Info($"Got {sequencedTransactions.Count} encrypted transactions from Shutter sequencer contract for slot {keys.Slot} at time {offsetText} slot start...");
+
+        Transaction[] transactions = DecryptSequencedTransactions(sequencedTransactions, keys.Keys);
+
+        if (_logger.IsDebug && transactions.Length > 0) _logger.Debug($"Decrypted Shutter transactions:{Environment.NewLine}{string.Join(Environment.NewLine, transactions.Select(tx => tx.ToShortString()))}");
+
+        Transaction[] filtered = FilterTransactions(transactions, parentHeader).ToArray();
```

Comment:
We just allocated few collections for transactions, can we reduce allocations here?
1. sequencedTransactions could be ArrayPoolList
2. transactions from decrypt could be IEnumerable

Reply:
I returned the decrypted transactions as a ParallelQuery as returning an IEnumerable would cause them to be decrypted sequentially. Also this means that filtering can happen in parallel